### PR TITLE
story(cmd/serve): read a config file

### LINKED ISSUE #27: story(cmd/serve): read a config file
### Description

As an end user, I want to be able to configure evrys through a config file, so that I can manage, track, and version that config file just like I would with any other code.

### Acceptance Criteria

- [ ] Persistent confg file flag is provided for all serve subcommands
- [ ] PersistentPreRun middleware is provided for reading config file

### Related Issues

_No response_

### PR DESCRIPTION
**Pull request by Zaba505**
Closes #27

### CODE REVIEW

File: cmd/cmd.go [hunk 1]
```diff
@@ -84,7 +86,65 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+	}
+}
+
+func initLogging(v *viper.Viper) func(*cobra.Command, []string) {
+	return func(cmd *cobra.Command, args []string) {
+		var lvl zapcore.Level
+		lvlStr := cmd.Flags().Lookup("log-level").Value.String()
+		err := lvl.UnmarshalText([]byte(lvlStr))
+		if err != nil {
+			panic(err)
```
Comment (erictg) [inline cmd/cmd.go:118]: Should exit with a code or return an error instead of panicking.

File: cmd/cmd.go [hunk 2]
```diff
@@ -84,7 +86,65 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+		}
+		if !flag.Changed {
+			return
+		}
+
+		v.SetConfigFile(flag.Value.String())
+		v.SetConfigType("yaml")
+		err := v.ReadInConfig()
+		if err != nil {
+			panic(err)
```
Comment (erictg) [inline cmd/cmd.go:147]: Same issue with panic

File: cmd/cmd_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+import (
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLoadConfigFile(t *testing.T) {
```
Comment (erictg) [inline cmd/cmd_test.go:25]: Panic fix will affect this test

**Review by erictg — CHANGES_REQUESTED** [review]
Other than the panics, pr is fine

File: cmd/cmd.go (hunk 1 above)
Comment (Zaba505) [reply cmd/cmd.go]: [a3fc9c1](https://github.com/z5labs/evrys/pull/33/commits/a3fc9c1cb25ad35ff2ba1530681c07ccb8998834)

File: cmd/cmd.go [hunk 4]
```diff
@@ -84,7 +86,65 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+		}
+
+		cfg := zap.NewProductionConfig()
+		cfg.Level = zap.NewAtomicLevelAt(zapcore.DebugLevel)
+		cfg.OutputPaths = []string{v.GetString("log-file")}
+		l, err := cfg.Build(zap.IncreaseLevel(lvl))
+		if err != nil {
+			panic(err)
```
[the remark this answers is not in the document]
Comment (Zaba505) [reply cmd/cmd.go]: [a3fc9c1](https://github.com/z5labs/evrys/pull/33/commits/a3fc9c1cb25ad35ff2ba1530681c07ccb8998834)

File: cmd/cmd.go (hunk 2 above)
Comment (Zaba505) [reply cmd/cmd.go]: [a3fc9c1](https://github.com/z5labs/evrys/pull/33/commits/a3fc9c1cb25ad35ff2ba1530681c07ccb8998834)

File: cmd/cmd_test.go (hunk 3 above)
Comment (Zaba505) [reply cmd/cmd_test.go:25]: [a3fc9c1](https://github.com/z5labs/evrys/pull/33/commits/a3fc9c1cb25ad35ff2ba1530681c07ccb8998834)

File: cmd/cmd.go [hunk 5]
```diff
@@ -84,7 +86,74 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+			bindFlags(v),
+			initLogging(v),
+		}
+		preRuns = append(preRuns, fs...)
+
+		return func(cmd *cobra.Command, args []string) error {
+			for _, f := range preRuns {
+				err := f(cmd, args)
+				if err != nil {
+					return err
```
Comment (erictg) [inline cmd/cmd.go:103]: error is not wrapped for context

File: cmd/cmd.go [hunk 6]
```diff
@@ -84,7 +86,74 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+	}
+}
+
+func initLogging(v *viper.Viper) preRunFunc {
+	return func(cmd *cobra.Command, args []string) error {
+		var lvl zapcore.Level
+		lvlStr := cmd.Flags().Lookup("log-level").Value.String()
+		err := lvl.UnmarshalText([]byte(lvlStr))
+		if err != nil {
+			return err
```
Comment (erictg) [inline cmd/cmd.go:125]: error is not wrapped for context

File: cmd/cmd.go [hunk 7]
```diff
@@ -84,7 +86,74 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+		}
+
+		cfg := zap.NewProductionConfig()
+		cfg.Level = zap.NewAtomicLevelAt(zapcore.DebugLevel)
+		cfg.OutputPaths = []string{v.GetString("log-file")}
+		l, err := cfg.Build(zap.IncreaseLevel(lvl))
+		if err != nil {
+			return err
```
Comment (erictg) [inline cmd/cmd.go:133]: error is not wrapped for context

File: cmd/cmd.go [hunk 8]
```diff
@@ -84,7 +86,74 @@ func buildCli(subcommandBuilders ...func(v *viper.Viper) *cobra.Command) *cobra.
...
+		}
+		if !flag.Changed {
+			return nil
+		}
+
+		v.SetConfigFile(flag.Value.String())
+		v.SetConfigType("yaml")
+		err := v.ReadInConfig()
+		if err != nil {
+			return err
```
Comment (erictg) [inline cmd/cmd.go:155]: error is not wrapped for context

**Review by erictg — CHANGES_REQUESTED** [review]
Errors need to be wrapped for context of what the error is from

File: cmd/cmd.go (hunk 8 above)
Comment (Zaba505) [reply cmd/cmd.go]: [456b83d](https://github.com/z5labs/evrys/pull/33/commits/456b83dd62b3f9033e24a9fe643c49b4629a5e9f)

File: cmd/cmd.go (hunk 7 above)
Comment (Zaba505) [reply cmd/cmd.go]: [456b83d](https://github.com/z5labs/evrys/pull/33/commits/456b83dd62b3f9033e24a9fe643c49b4629a5e9f)

File: cmd/cmd.go (hunk 6 above)
Comment (Zaba505) [reply cmd/cmd.go]: [456b83d](https://github.com/z5labs/evrys/pull/33/commits/456b83dd62b3f9033e24a9fe643c49b4629a5e9f)

File: cmd/cmd.go (hunk 5 above)
Comment (Zaba505) [reply cmd/cmd.go]: [456b83d](https://github.com/z5labs/evrys/pull/33/commits/456b83dd62b3f9033e24a9fe643c49b4629a5e9f)

**Review by erictg — APPROVED** [review]

**Pull request merged**